MANUAL.md

```diff
@@ -0,0 +1,194 @@
+        self.post(
+            ApiKind::Service("my_service"),
+            "v1/third_endpoint",
+            &data,
+        )
+    }
+}
+
+// Api test skeleton
+
+#[test]
+fn my_api_test() {
+    let mut testkit = TestKitBuilder::validator()
+        .with_service(MyService::new())
+        .create();
+    fill_storage_with_data(&mut testkit);
+    // Check api responses
+    let api = testkit.api();
+    assert_eq!(api.get_public_data(), ApiResponsePublicData::new(...));
+    ...
+}
+```
+
+## Advanced usage
+
+Here are examples of more complex and less common cases.
+
+### Oracles testing
+
+The oracle in this case is service which can produce transaction with external data after the commit of block,
+[`exonum-time`][exonum-time] and [`exonum-btc-anchoring`][exonum-btc-anchoring] are examples of these kind of oracles.
+In this way, transactions which were created during the `handle_commit` execution will be stored in `TestKit` memory pool.
+
+```rust
+let mut testkit = TestKitBuilder::validator()
+    .with_service(HandleCommitService)
+    .create();
+// Check that `handle_commit` invoked on the correct height.
+for i in 1..5 {
+    testkit.create_block();
+    let tx = TxAfterCommit::new_with_signature(Height(i), &Signature::zero());
+    assert!(testkit.mempool().contains_key(&tx.hash()));
+}
+```
+
+In order to invoke a `handle_commit` event, you must create a block.
+If the oracle has to fetch any data from external world, you must create a mock object.
+
+```rust
+// Provide a mock api for the service.
+let mut cruel_world = ExternalApiMock::new();
+let mut testkit = TestKitBuilder::validator()
+    .with_service(MyOracleService::with_client(cruel_world.client()))
+    .create();
+// Expect the request from the service.
+cruel_world.expect_api_call(ApiCallInfo { ... })
+    .with_response_ok(ApiResponse { ... });
+// Call the `handle_commit` event.
+testkit.create_block();
+let expected_tx = MyOracleTx::new(...);
+// Check that the expected transaction there is in memory pool.
+assert!(testkit.mempool().contains_key(&expected_tx.hash()));
+```
+
+### Configuration changes testing
+
+If your service has its own configuration, you may need to test the response to a configuration change.
+With the testkit you can create the configuration change proposal and commit them.
+
+```rust
+// Provide a mock api for the service.
+let mut testkit = TestKitBuilder::validator()
+    .with_service(MyOracleService::new())
+    .create();
+// Create configuration change proposal.
+let proposal = {
+    let mut cfg = testkit.configuration_change_proposal();
+    cfg.set_actual_from(cfg_change_height);
+    cfg.set_service_config("my_service", MyServiceCfg { ... });
+    cfg
+};
+let stored = proposal.stored_configuration().clone();
+testkit.commit_configuration_change(proposal);
+// Check that following configuration is none.
+assert_eq!(
+    Schema::new(&testkit.snapshot()).following_configuration(),
```

Comment:
Nitpick: maybe, write an explicit `use` statement for importing `Schema` into the scope? On its own, `Schema` seems quite generic, so it could be mistaken for the schema of some service.

Reply:
Sounds reasonable.